integration/simple-visitorPattern/simple.ts

```diff
@@ -0,0 +1,224 @@
+/* eslint-disable */
+import * as _m0 from "protobufjs/minimal";
+
+export const protobufPackage = "simple";
+
+export interface PBVisitable {
+  visit(visitor: Visitor): void;
```

Comment:
The philosophy in it's core is conflicting with the Visitor pattern that is an object oriented pattern that uses indirection based on classes. Could you elaborate your idea on how this would function?

Reply:
Combining the two comment threads...

> it is up to the visitor to decide what are the interesting parts of the message

Imo that is not traditionally what the visitor pattern is, i.e. see the [wikipedia](https://en.wikipedia.org/wiki/Visitor_pattern) implementations:

```java
    @Override
    public void accept(CarElementVisitor visitor) {
        for (CarElement element : elements) {
            element.accept(visitor);
        }
        visitor.visit(this);
    }
```

```python
class Car(CarElement):
    def __init__(self):
        self.elements = [
            Wheel("front left"), Wheel("front right"),
            Wheel("back left"), Wheel("back right"),
            Body(), Engine()
        ]

    def accept(self, visitor):
        for element in self.elements:
            element.accept(visitor)
        visitor.visitCar(self)
```

They all have the parent's `accept` method auto-traversing into the child by auto-calling the child `accept`s within it's own `accept` method.

That is what the visitor pattern fundamentally is, b/c the idea comes from compilers/ASTs/etc. where the number/complexity/types of children is large, non-trivial, and would be easy for consumers to forget/not get right, so the visitor impl handles auto-traversal.

I get it reduces control, and b/c of that visitors can just be complicated to use in practice, but :shrug: afaiu that is just pros/cons of the pattern.

To make it concrete in terms of this PR, if all `accept` does is call `visitor.visitWhateverIAm()`, that's so simple, couldn't this code:

```ts
  visitSimple(obj: Simple): void {
    this.simpleTree = this.simpleTree + obj.name + '\r\n';
    for (const child of obj.child) {
      child.accept(this);
    }
  }
```

Just be written as:

```ts
  visitSimple(obj: Simple): void {
    this.simpleTree = this.simpleTree + obj.name + '\r\n';
    for (const child of obj.child) {
      this.visitChild(child);
    }
  }
```

Where I just manually changed `child.accept` --> `this.visitChild(child)`. If that is all that `child.accept` is going to do, I don't think it's carrying it's weight as an abstraction, so you might as well just call `this.visitChild(child)` directly, and it would work without any changes to ts-proto's output.

Personally, given you explicitly want to control iteration, that's my recommendation for what you should do. That said, if you really want to see visitors-the-pattern implemented in ts-proto, that's great too and we can keep discussing, but I think, for me, having the visitor control iteration will be a required requirement, b/c again w/o that, why bother/you don't really need ts-proto's help to call `child.accept(this)` instead of `this.visitChild(child)`.

> The philosophy in it's core is conflicting with the Visitor pattern that is
> an object oriented pattern that uses indirection based on classes.

I agree that the visitor pattern is typically implemented as a polymorphic `accept` method, but that is really only required for situations that have _hetergenous_ collections, i.e. something like:

```
class FruitBasket {
  fruits = [new Apple(), new Banana()]
  accept(visitor) {
    // this is heterogeneous, we don't know to call f.acceptApple or f.acceptBanana
    // so we just call f.accept and let OO figure it out
    this.fruits.forEach(f => f.accept(visitor))
  }
}
```

However, protobuf fundamentally does not support heterogeneous collections; when we have a repeated list of Apples in a message, we know all of them are Apples. When we have a repeated list of Bananas in a message, we know all of them are Bananas. Because of this, we don't have the "I don't know whether to call `acceptApple` or `acceptBanana`, and so our generated code can "just know" which static method (Accept.accept or Banana.accept) to call and it will always be correct.